pkg/azure/types/types.go

```diff
@@ -113,7 +113,7 @@ type AzureInterface struct {
 	// GatewayIP is the interface subnet's default route
 	//
 	// +optional
-	GatewayIP string `json:"-"`
+	GatewayIP string `json:"gateway-ip"`
```

Comment:
https://github.com/cilium/cilium/commit/8de6dc5a02b1af20d95bcc9c3783c9b08b2157c0 did break the API by adding the `json:"-"` tag.

Unfortunately, to not break the API, 
```
GatewayIP string `json:"gateway-ip"`
```
needs to be
```
GatewayIP string `json:"GatewayIP"`
```
as it was introduced in https://github.com/cilium/cilium/pull/11268/files#diff-54f340ec296f0270b40d38b4245bbb130a622e4b05315285f98b29629a4e0f31R116

Perhaps it is better to keep the `json:"GatewayIP"`, backport it to 1.9 and in 1.10 add a new field (`json:"gateway-ip"`) that would allow us to have consistency across all json tags in `AzureInterface` while deprecating `json:"GatewayIP"` and removing it in 1.11.

Reply:
@aanm, I don't understand how my change will break the API because it was introduced in https://github.com/cilium/cilium/pull/11268/files#diff-54f340ec296f0270b40d38b4245bbb130a622e4b05315285f98b29629a4e0f31R116. I'm adding a json tag whereas the code you're referring to seems to be using the struct field.

If I understand the codebase correctly, then this is what is happening - 
1. `cilium-operator-azure` will update the `status` of `CiliumNode` CR with `GatewayIP` information [here](https://github.com/cilium/cilium/blob/769123365c08746eb228fbd231b0cc1833eacca0/pkg/ipam/node.go#L784). But I think kube-apiserver ignores it as `GatewayIP` is not present in the CRD schema. 
2. `cilium-agent` tries to fetch `GatewayIP` of the AzureInterface from CiliumNode CR but since it is not present in the CRD schema, it will be empty. 

With my change which adds a `gateway-ip` field in CRD schema,
1. `cilium-operator-azure` will be successful in populating `gateway-ip` field in `CiliumNode` CR.
2. `cilium-agent` will be successful in fetching the `GatewayIP` of the AzureInterface as it is populated in `CiliumNode` CR.

I tested my changes in 1.18 Azure cluster and it seems to be working. I ensured that there is per-NIC routing table using this command `ip rule`. There was no routing table for the secondary NIC before my changes.

Reply:
@AnishShah when the field was introduced in https://github.com/cilium/cilium/pull/11268/files#diff-54f340ec296f0270b40d38b4245bbb130a622e4b05315285f98b29629a4e0f31R116 the JSON key of that field was `GatewayIP`. This PR is changing the API form `GatewayIP` to `gateway-ip`.

Reply:
Ok. Since this is a pull request against master, I'll add one commit which changes the JSON tag to `GatewayIP`. Another commit which deprecates this field and adds another field. Any idea what should I name the new field? Just `Gateway` or `GatewayAddress`

Reply:
Also I'm not sure how I can make a struct field as deprecated. Any examples on how to do it?

Reply:
We usually just mark the field as deprecated by putting a comment on the field saying so: https://github.com/cilium/cilium/blob/ffd02dd37aebbea366df9cadc752fe95fb2ba137/pkg/aws/eni/types/types.go#L49

Reply:
Fixed.